lib/fast-path.ts

```diff
@@ -305,52 +305,43 @@ FPp.needsParens = function(assumeExpressionContext) {
     return false;
   }
 
-  const name = this.getName();
-
   // If the value of this path is some child of a Node and not a Node
   // itself, then it doesn't need parentheses. Only Node objects (in fact,
   // only Expression nodes) need parentheses.
   if (this.getValue() !== node) {
     return false;
   }
 
-  // Only statements don't need parentheses.
-  if (n.Statement.check(node)) {
+  if (parent.type === "ParenthesizedExpression" || (node.extra && node.extra.parenthesized)) {
     return false;
   }
 
+  switch (node.type) {
+  // Only statements don't need parentheses.
+  case "Statement":
+    return false;
   // Identifiers never need parentheses.
-  if (node.type === "Identifier") {
+  case "Identifier":
     return false;
-  }
 
-  if (parent.type === "ParenthesizedExpression") {
-    return false;
-  }
-
-  switch (node.type) {
   case "UnaryExpression":
   case "SpreadElement":
   case "SpreadProperty":
-    return parent.type === "MemberExpression"
-      && name === "object"
-      && parent.object === node;
```

Comment:
Both of these checks do exactly the same thing. I eliminated the duplication and replaced all the name checks with reference equality checks against node.

Reply:
They don't exactly always mean the same thing. You could have a node that's reused for more than one child, so `parent.property === node`, but the name of the child along the current traversal path is not actually `property`, because we're printing a different child right now. This is an obscure point for sure, but I'd like to keep the existing (perhaps excessively careful) logic.